### PR TITLE
Add initial benchmarks

### PR DESCRIPTION
**Pull request by davidpdrsn**
This adds a basic setup for writing benchmarks and adds a couple of benchmarks.

The output looks like this

```
❯ cargo bench -p axum -- send-json
   Compiling axum v0.5.10
    Finished bench [optimized] target(s) in 0.68s
     Running unittests src/lib.rs (target/release/deps/axum-732f709dd139ce93)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 203 filtered out; finished in 0.00s

     Running benches/benches.rs (target/release/deps/benches-ddcccec176823e7f)
Running "send-json" benchmark
  Beginning round 1...
  Benchmarking 10 connections @ http://0.0.0.0:58714 for 10 second(s)
    Latencies:
      Avg      Stdev    Min      Max
      0.06ms   0.03ms   0.02ms   4.06ms
    Requests:
      Total: 1689726 Req/Sec: 168960.39
    Transfer:
      Total: 228.83 MB Transfer Rate: 22.88 MB/Sec
```

It basically runs a default hyper server and shells out to [rewrk](https://github.com/lnx-search/rewrk) which does all the heavy lifting.

I went with rewrk rather than the default cargo bench tooling (or criterion) because I'm mainly interested in seeing requests per second and not how long a single request takes. I would also like to avoid configuring and running our own client. This seems fine to be me but let me know what you think.

Fixes https://github.com/tokio-rs/axum/issues/1199

### CODE REVIEW

File: axum/benches/benches.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    fn run<F>(self, f: F)
+    where
+        F: FnOnce() -> Router,
+    {
+        // support only running some benchmarks with
+        // ```
+        // cargo bench -- routing send-json
+        // ```
+        let args = std::env::args().collect::<Vec<_>>();
+        let names = &args[1..args.len() - 1];
```
Comment (jplatte) [inline axum/benches/benches.rs:100]: So wasteful to collect the commandline args once per call of `run` 😭
At least we can do slightly less work w/o complicating the code:

```suggestion
        let names = std::env::args().skip(1).collect::<Vec<_>>();
```

File: axum/benches/benches.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+        let mut cmd = Command::new("rewrk");
+        cmd.stdout(Stdio::piped());
+
+        cmd.arg("--host");
+        if let Some(path) = self.path {
+            cmd.arg(format!("http://{}{}", addr, path));
+        } else {
+            cmd.arg(format!("http://{}", addr));
+        }
```
Comment (jplatte) [inline axum/benches/benches.rs:135]: Could slightly simplify to `format!("http://{}{}, addr, path.unwrap_or(""))`.

File: axum/benches/benches.rs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+        for (key, value) in self.headers.into_iter().flatten() {
+            cmd.arg("--header");
+            cmd.arg(format!("{}: {}", key, value));
+        }
+
+        if let Some(body) = self.body {
+            cmd.arg("--body");
+            cmd.arg(body);
```
Comment (jplatte) [inline axum/benches/benches.rs:153]:
```suggestion
            cmd.args(&["--body", body]);
```

**Comment by jplatte** [comment]
Overall looks very decent :)

**Comment by programatik29** [comment]
I observed rewrk doesn't exit with error code on error.
Errors can be detected by checking stderr.

**Comment by davidpdrsn** [comment]
> I observed rewrk doesn't exit with error code on error.
> 
> Errors can be detected by checking stderr.

We should fix that. Checking stderr is tedious 😅

**Review by jplatte — APPROVED** [review]
The bulk of this looks good.

**Pull request merged**